### ISSUE TITLE
Loud failures if donate_argnums does not work?

### ISSUE THREAD

**Issue by girving** [jax-ml/jax#9675]
Currently jax provides a gentle warning if a donated buffer in jit is not reused:

https://github.com/google/jax/blob/ab15db7d8658825afa9709df46f0dea76688d309/jax/_src/api.py#L913

Is it possible to add make this optionally a fatal error throwing an exception, to avoid subtle memory regressions in code using donation?  One option would be an extra boolean argument to `jax.jit`, though admittedly that's a bit inelegant.

Another option (suggested by Malcolm Reynolds) would be a `jax.config` option that changes this globally.

**Comment by rajasekharporeddy** [comment]
Hi @girving 

To confirm my understanding, regarding the behavior of donated buffers in JAX, are you expecting JAX to raise an exception or error message, when a donated buffer is reused? If so, JAX now raises a `ValueError: INVALID_ARGUMENT: Invalid buffer passed: buffer has been deleted or donated.`, when tried to reuse a donated buffer.

```python
import jax
import jax.numpy as jnp

print("Jax Version: ", jax.__version__)

buf = jax.numpy.array(1)

def f(x):
  return x+1

f_jit = jax.jit(f, donate_argnums=(0,))

f_jit(buf)
f_jit(buf)
```
Output
```python
Jax Version:  0.5.2
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-2-c8e53a5c6834>](https://localhost:8080/#) in <cell line: 0>()
     12 
     13 f_jit(buf)
---> 14 f_jit(buf)

ValueError: INVALID_ARGUMENT: Invalid buffer passed: buffer has been deleted or donated.
```
Attaching the [gist](https://colab.sandbox.google.com/gist/rajasekharporeddy/c92334a02e39782f3d83a975a8bdd0b5/-9675.ipynb) for reference. Thank you.

**Comment by girving** [comment]
Yes, I’d want it to raise an exception. Good progress if it does now (3 years later :))!

**Comment by rajasekharporeddy** [comment]
Please feel free to close the issue, if resolved.
